File: dailyQuestion/2020/2020-06/06-15/python/solution_dfs.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """

        def rserialize(root, s):
            if not root:
                s += "null,"
                return s
            s += "{},".format(root.val)
            s = rserialize(root.left, s)
            s = rserialize(root.right, s)
            return s

        return rserialize(root, "")

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        l = []
        _tmp = data.split(",")
        for item in _tmp:
            if item:
                l.append(item)
        l.reverse()

        def rdeserialize():
            if not l:
                return None
            val = l.pop()
            if val == "null":
                return None
            root = TreeNode(val)
            root.left = rdeserialize()
            root.right = rdeserialize()
            return root

        return rdeserialize()

if __name__ == "__main__" :
    codec = Codec()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)

    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)

    data = codec.serialize(root)
    result = codec.deserialize(data)
    print(result)

File: dailyQuestion/2020/2020-06/06-15/golang/solution_dfs.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	obj := Constructor()
	data := "1,2,3,null,null,4,5"
	ans := obj.deserialize(data)
	data1 := obj.serialize(ans)
	fmt.Println(data1)
	fmt.Println(ans)

}

type Codec struct {
	list []string
}

func Constructor() Codec {
	return Codec{}
}

func (this *Codec) rserialize(root *TreeNode, str string) string {
	if root == nil {
		str += "null,"
	} else {
		str += strconv.Itoa(root.Val) + ","
		str = this.rserialize(root.Left, str)
		str = this.rserialize(root.Right, str)
	}
	return str
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	return this.rserialize(root, "")
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	list := strings.Split(data, ",")
	for i := 0; i < len(list); i++ {
		if list[i] != "" {
			this.list = append(this.list, list[i])
		}
	}
	return this.rdeserialize()
}

func (this *Codec) rdeserialize() *TreeNode {
	if len(this.list) == 0 {
		return nil
	}
	if this.list[0] == "null" {
		this.list = this.list[1:]
		return nil
	}
	val, _ := strconv.Atoi(this.list[0])
	this.list = this.list[1:]
	root := &TreeNode{Val: val}
	root.Left = this.rdeserialize()
	root.Right = this.rdeserialize()
	return root
}
